File: homework/hw3/hw3.py
"""Code for M3C 2016 Homework 3 Mathilde Duverger CID:00978498
"""
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from n1 import network as net 

def visualize(e):    
    """Visualize a network represented by the edge list
    contained in e
    """
    n = max(np.unique(e))
    fig, ax = plt.subplots()
    X = []
    Y = []
    #plots the nodes
    for i in range(n):
        x = np.random.uniform()
        y = np.random.uniform()
        X.extend([x])
        Y.extend([y])
        circle = plt.Circle((x,y),0.01) 
        ax.add_artist(circle)
    #plots the lines between the nodes
    for i in range(np.shape(e)[0]):
        line = matplotlib.lines.Line2D((X[e[i,0]-1],X[e[i,1]-1]),(Y[e[i,0]-1],Y[e[i,1]-1]),linewidth=0.2,color='black')
        ax.add_line(line)
    plt.show()
 
def degree(N0,L,Nt,display=False):
    """Compute and analyze degree distribution based on degree list, q,
    corresponding to a recursive network (with model parameters specified
    as input.
    The functional form seen in the saved figure is the exponential of 
    the polyfit of degree 3 to the log of the first Np/2 values of P that
    are non zero.
    """
    qmax = net.generate(N0,L,Nt)
    q = net.qnet.tolist()
    P = np.zeros(qmax)
    d = np.array(range(1,qmax+1))
    #compute degree distribution
    for i in d:
        P[i-1] = float(q.count(i))/float(np.size(q))
    Np = np.size(P)
    #if Nt>1000 compute a functional form for P(d)
    if (Nt>1000):
        P1 = (P[:Np/2 +1])
        P1 = P1[np.nonzero(P1)]
        lP1 = np.log(P1)
        a1,a2,a3,a4 = np.polyfit(d[np.nonzero(P1)],lP1,3)
        f = np.exp(a1*np.power(d,3)+a2*np.power(d,2)+a3*np.power(d,1)+a4*np.power(d,0))
    #if display is True then plot the distribution and fit if Nt>1000
    if (display==True):
        width = 1/1.5
        plt.figure()
        plt.bar(d,P,width,align='center')
        if (Nt>1000):
            plt.plot(d,f,'r',label='Fit to P(d)')
        plt.xticks(np.arange(min(d), max(d)+1, 10))
        plt.title('Mathilde Duverger degree')
        plt.xlabel('d')
        plt.ylabel('P(d)')
        plt.legend(loc='best')
        plt.show()
    return d,P
    
#if __name__ == "__main__":

File: homework/hw3/hw3.f90
!Code for M3C 2016 Homework 3 Mathilde Duverger CID:00978498
module network
    implicit none
    integer, dimension(:), allocatable :: qnet
    integer, dimension(:,:), allocatable :: anet,enet 
    save 
    contains 

subroutine generate(N0,L,Nt,qmax)
	!Generate recursive matrix corresponding
	!to model parameters provided as input
	implicit none
	integer, intent(in) :: N0,L,Nt
	integer, intent(out) :: qmax
        real(kind=8), dimension(:), allocatable :: PQ,P,L1
	integer :: i,j,t
	!initialise model with N0 nodes 
	allocate ( qnet(N0+Nt) )
	allocate( enet(2,N0+Nt*L) ) 
	do i = 1,N0-1
	   qnet(i) = 2
	   enet(1,i) = i
	   enet(2,i) = i+1
	end do
        qnet(N0) = 2
        enet(1,N0) = N0
        enet(2,N0) = 1
        do i = N0+1,N0+Nt
            qnet(i) = 0
        end do
        !run model for t from 1 to Nt
        do t = 1,Nt
            !generate the probabilities vector PQ
            allocate( PQ(N0+t-1) )
            do i =1,N0+t-1
                PQ(i) = qnet(i)/dble(sum(qnet))
            end do
            !generate the cmf vector P
            allocate( P(N0+t-1) )
            do i=1,N0+t-1
                P(i) = sum(PQ(1:i))
            end do
            !generate a random sample of size L from 
            !the uniform distribution
            allocate( L1(L) )
            call random_number(L1)
            !for each element in L1 assign it to
            !a certain node if the probability at 
            !L1(j) is smaller than the cmf for 
            !that node
            do i =1,L
                do j = 1,size(P)
                    if (L1(i)<=P(j)) then
                        L1(i) = j
                        exit  
                    end if
                end do
            end do
            !add the new links created to the degrees of each node
            do i = 1,L
                j = L1(i)
                qnet(j) = qnet(j) +1
            end do 
            qnet(N0 + t) = L
            !add the new links created to our edge list
            do i = 1,L
                enet(1,N0+(t-1)*L+i) = N0+t 
                j = L1(i)
                enet(2,N0+(t-1)*L+i) = j
            end do
            deallocate( PQ )
            deallocate( P )
            deallocate( L1 ) 
        end do
        !find degree of largest hub in network
        qmax = maxval(qnet)
end subroutine generate


subroutine adjacency_matrix()
	!Create adjacency matrix corresponding 
	!to pre-existing edge list 
	implicit none
	integer :: i
	allocate( anet(size(qnet),size(qnet)) )
        do i =1,size(enet)/2
            anet(enet(1,i),enet(2,i))=1
            anet(enet(2,i),enet(1,i))=1
        end do


end subroutine adjacency_matrix

subroutine connectivity(c)
	!Compute connectivity of pre-computed network
	implicit none
	real(kind=8), intent(out) :: c
	integer, dimension(:,:), allocatable :: M,D
	integer, dimension(2) :: S
	integer :: i,LWORK,INFO
	external :: DSYEV
        intrinsic :: INT, MIN
        double precision, dimension(:,:),allocatable :: Mtemp
        double precision, dimension(:),allocatable :: WORK
        double precision, dimension(:), allocatable :: W
        S = shape(anet)
        allocate( M(S(1),S(2)))
        allocate( D(S(1),S(2)))
        !compute M the Laplacian matrix
        do i =1,size(qnet)
            D(i,i) = qnet(i)
        end do
        M = D - anet
        
        !run DSYEV
        allocate( Mtemp(S(1),S(2)))
        Mtemp = dble(M)
        allocate( WORK(1) )
        allocate( W(S(1)) )
        !query the optimal workspace
        LWORK = -1
        CALL DSYEV( 'N', 'Upper', S(1), Mtemp, S(1), W, WORK, LWORK, INFO )
        LWORK = WORK(1)
        deallocate( WORK )
        !solve eigenproblem 
        allocate( WORK(MAX(1,LWORK)) )
        CALL DSYEV( 'N', 'Upper', S(1), Mtemp, S(1), W, WORK, LWORK, INFO )
        !W is the array of eigenvalues in ascending order, 
        !c is the second smallest eigenvalue
        c = W(2)
end subroutine connectivity


subroutine vary_connectivity(N0,Lmax,Nt,carray)
	!Compute connectivity for networks with L varying
	!between 1 and Lmax
	!As the number of links added L increases in every loop
	!the connectivity decreases. There is a linear correspondence 
	!between L and c.
	implicit none
	integer, intent(in) :: N0,Lmax,Nt
	real(kind=8), dimension(Lmax),intent(out) :: carray
	integer :: L,qmax
	real(kind=8) :: c
        do L =1,Lmax
            call generate(N0,L,Nt,qmax)
            call adjacency_matrix() 
            call connectivity(c)
            carray(L) = c
            deallocate( qnet )
            deallocate( enet )
            deallocate( anet )
        end do
end subroutine vary_connectivity

end module network
